### PR TITLE
ui: use stencil buffer to draw polygons

### PR DESCRIPTION
**Pull request by sshane**
When drawing very thin lane lines and you set a brush, UI frame times shoot up to hundreds of ms. If you don't draw the polygon, or don't set a brush, there's no issues. I tried to disable anti-aliasing, but that had no effect. The following plots are just with one lane line for isolation. As llp[0] nears 0, the frame times get worse and worse. @deanlee can you think of any reason why this happens and a better way to fix it instead of removing the dynamic width?

https://github.com/commaai/openpilot/blob/80eff1b6ce6600c8b93846ac6e096cf38f9ec695/selfdrive/ui/qt/onroad.cc#L444-L448

https://github.com/commaai/openpilot/assets/25857203/36c9261b-3bd7-48d6-8adb-21cc841d7abf

image

image

### CODE REVIEW

File: selfdrive/ui/ui.cc [hunk 1]
```diff
@@ -96,7 +96,7 @@ void update_model(UIState *s,
   int max_idx = get_path_length_idx(lane_lines[0], max_distance);
   for (int i = 0; i < std::size(scene.lane_line_vertices); i++) {
     scene.lane_line_probs[i] = lane_line_probs[i];
-    update_line_data(s, lane_lines[i], 0.025 * scene.lane_line_probs[i], 0, &scene.lane_line_vertices[i], max_idx);
+    update_line_data(s, lane_lines[i], 0.025, 0, &scene.lane_line_vertices[i], max_idx);
```
Comment (sshane) [inline selfdrive/ui/ui.cc:99]:
```suggestion
    update_line_data(s, lane_lines[i], 0.025 * fmin(scene.lane_line_probs[i], 0.5), 0, &scene.lane_line_vertices[i], max_idx);
```

it looks pretty ugly without something, maybe clipping to half is fine.

before:

image

after:

image

**Comment by deanlee** [comment]
I had previously submitted a PR #23542 to not draw lane line when the llp nears 0. However, I don't know why it behaves this way. While changing brushes frequently can impact performance, especially when drawing on an OpenGL surface, it shouldn't be so noticeable.

**Comment by sshane** [comment]
I was able to reproduce it when I only set one brush (to white) and drew one lane line

**Comment by deanlee** [comment]
Will there be an improvement if "setSamples"  to a lower value? current value is 16:

https://github.com/commaai/openpilot/blob/d68bd905fc9440352007e115c0f46dc9c0b4e402/selfdrive/ui/qt/util.cc#L98

**Comment by sshane** [comment]
No, I even disabled it with `-1`, same issue. I can even reproduce it on PC if you want the rlog

**Comment by sshane** [comment]
Ah, max_distance from the plan is almost always above 100 when testing, but model is much less. If you hardcode max_distance to 99, the spikes mostly go away. any ideas @deanlee?

**Comment by sshane** [comment]
Certain QPolygonFs cause these high frame times consistently, for example this waits until a high draw time, then keeps drawing it:

```
selfdrive/ui/ui.cc: max_distance orig, new: 78.3472 100
selfdrive/ui/qt/onroad.cc: using bad poly
selfdrive/ui/qt/onroad.cc: is closed: false
selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-394.075,900.027)QPointF(-54.5725,810.538)QPointF(180.188,748.413)QPointF(341.522,705.421)QPointF(467.642,671.917)QPointF(563.719,647.199)QPointF(624.822,628.565)QPointF(675.679,613.808)QPointF(702.328,601.059)QPointF(725.26,591.128)QPointF(733.876,585.022)QPointF(754.121,577.493)QPointF(751.813,572.547)QPointF(766.213,569.503)QPointF(771.051,564.169)QPointF(762.17,565.01)QPointF(775.544,562.746)QPointF(775.581,562.746)QPointF(762.21,565.01)QPointF(771.095,564.169)QPointF(766.261,569.503)QPointF(751.866,572.547)QPointF(754.181,577.493)QPointF(733.944,585.022)QPointF(725.336,591.128)QPointF(702.414,601.059)QPointF(675.778,613.808)QPointF(624.937,628.565)QPointF(563.854,647.199)QPointF(467.803,671.917)QPointF(341.716,705.421)QPointF(180.428,748.413)QPointF(-54.2691,810.538)QPointF(-393.679,900.027))
selfdrive/ui/qt/onroad.cc: draw time: 1.05996 ms  

selfdrive/ui/ui.cc: max_distance orig, new: 66.7283 100
selfdrive/ui/qt/onroad.cc: using bad poly
selfdrive/ui/qt/onroad.cc: is closed: false
selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-310.599,902.3)QPointF(4.31749,814.123)QPointF(224.051,753.363)QPointF(368.411,712.633)QPointF(486.291,680.94)QPointF(580.094,657.547)QPointF(625.66,640.1)QPointF(682.804,625.856)QPointF(702.245,613.073)QPointF(725.857,603.528)QPointF(735.763,597.09)QPointF(754.062,589.918)QPointF(750.446,584.782)QPointF(766.495,580.99)QPointF(772.2,575.44)QPointF(759.51,576.265)QPointF(770.563,574.463)QPointF(770.599,574.463)QPointF(759.55,576.265)QPointF(772.244,575.44)QPointF(766.543,580.99)QPointF(750.5,584.782)QPointF(754.122,589.918)QPointF(735.83,597.09)QPointF(725.933,603.528)QPointF(702.331,613.073)QPointF(682.903,625.856)QPointF(625.775,640.1)QPointF(580.229,657.547)QPointF(486.451,680.94)QPointF(368.606,712.633)QPointF(224.29,753.363)QPointF(4.62096,814.123)QPointF(-310.203,902.3))
selfdrive/ui/qt/onroad.cc: draw time: 18.168 ms  

selfdrive/ui/ui.cc: max_distance orig, new: 73.8285 100
selfdrive/ui/qt/onroad.cc: using bad poly
selfdrive/ui/qt/onroad.cc: is closed: false
selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-451.948,899.583)QPointF(-99.3411,810.203)QPointF(146.246,748.129)QPointF(319.808,704.831)QPointF(455.683,670.959)QPointF(564.931,645.632)QPointF(635.444,626.562)QPointF(697.379,611.425)QPointF(735.719,598.36)QPointF(768.944,587.918)QPointF(786.11,581.553)QPointF(816.362,573.425)QPointF(821.051,568.269)QPointF(843.466,564.851)QPointF(852.994,559.49)QPointF(848.498,560.107)QPointF(864.489,558.175)QPointF(864.526,558.175)QPointF(848.538,560.107)QPointF(853.038,559.49)QPointF(843.514,564.851)QPointF(821.105,568.269)QPointF(816.422,573.425)QPointF(786.177,581.553)QPointF(769.019,587.918)QPointF(735.806,598.36)QPointF(697.478,611.425)QPointF(635.559,626.562)QPointF(565.065,645.632)QPointF(455.843,670.959)QPointF(320.003,704.831)QPointF(146.485,748.129)QPointF(-99.0377,810.203)QPointF(-451.552,899.583))
selfdrive/ui/qt/onroad.cc: draw time: 1.18842 ms  

selfdrive/ui/ui.cc: max_distance orig, new: 100 100
selfdrive/ui/qt/onroad.cc: using bad poly
selfdrive/ui/qt/onroad.cc: is closed: false
selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-421.356,904.149)QPointF(-78.3551,813.953)QPointF(154.816,751.289)QPointF(318.861,707.915)QPointF(442.034,674.219)QPointF(540.446,649.28)QPointF(606.151,630.116)QPointF(655.488,615.075)QPointF(687.746,601.825)QPointF(713.056,591.03)QPointF(723.581,584.804)QPointF(747.622,576.588)QPointF(747.899,571.431)QPointF(766.482,568.15)QPointF(774.357,562.801)QPointF(767.935,563.462)QPointF(785.301,560.92)QPointF(785.337,560.92)QPointF(767.975,563.462)QPointF(774.401,562.801)QPointF(766.53,568.15)QPointF(747.953,571.431)QPointF(747.682,576.588)QPointF(723.649,584.804)QPointF(713.132,591.03)QPointF(687.833,601.825)QPointF(655.588,615.075)QPointF(606.266,630.116)QPointF(540.581,649.28)QPointF(442.195,674.219)QPointF(319.055,707.915)QPointF(155.056,751.289)QPointF(-78.0515,813.953)QPointF(-420.96,904.149))
selfdrive/ui/qt/onroad.cc: draw time: 1.28275 ms  

selfdrive/ui/ui.cc: max_distance orig, new: 86.5993 100
selfdrive/ui/qt/onroad.cc: using bad poly
selfdrive/ui/qt/onroad.cc: is closed: false
selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-227.194,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
selfdrive/ui/qt/onroad.cc: draw time: 54.2867 ms high! 

selfdrive/ui/ui.cc: max_distance orig, new: 82.4211 100
selfdrive/ui/qt/onroad.cc: is closed: true
selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-226.891,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
selfdrive/ui/qt/onroad.cc: draw time: 48.1844 ms high! 

selfdrive/ui/ui.cc: max_distance orig, new: 61.5824 100
selfdrive/ui/qt/onroad.cc: is closed: true
selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-226.891,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
selfdrive/ui/qt/onroad.cc: draw time: 48.1316 ms high!
```

**Comment by deanlee** [comment]
> Certain QPolygonFs cause these high frame times consistently, for example this waits until a high draw time, then keeps drawing it. It then is fixed if you convert it to an integer polygon with .toPolygon()
> 
> ```
> selfdrive/ui/ui.cc: max_distance orig, new: 73.8285 100
> selfdrive/ui/qt/onroad.cc: is closed: false
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-451.948,899.583)QPointF(-99.3411,810.203)QPointF(146.246,748.129)QPointF(319.808,704.831)QPointF(455.683,670.959)QPointF(564.931,645.632)QPointF(635.444,626.562)QPointF(697.379,611.425)QPointF(735.719,598.36)QPointF(768.944,587.918)QPointF(786.11,581.553)QPointF(816.362,573.425)QPointF(821.051,568.269)QPointF(843.466,564.851)QPointF(852.994,559.49)QPointF(848.498,560.107)QPointF(864.489,558.175)QPointF(864.526,558.175)QPointF(848.538,560.107)QPointF(853.038,559.49)QPointF(843.514,564.851)QPointF(821.105,568.269)QPointF(816.422,573.425)QPointF(786.177,581.553)QPointF(769.019,587.918)QPointF(735.806,598.36)QPointF(697.478,611.425)QPointF(635.559,626.562)QPointF(565.065,645.632)QPointF(455.843,670.959)QPointF(320.003,704.831)QPointF(146.485,748.129)QPointF(-99.0377,810.203)QPointF(-451.552,899.583))
> selfdrive/ui/qt/onroad.cc: draw time: 1.18842 ms  
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 100 100
> selfdrive/ui/qt/onroad.cc: is closed: false
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-421.356,904.149)QPointF(-78.3551,813.953)QPointF(154.816,751.289)QPointF(318.861,707.915)QPointF(442.034,674.219)QPointF(540.446,649.28)QPointF(606.151,630.116)QPointF(655.488,615.075)QPointF(687.746,601.825)QPointF(713.056,591.03)QPointF(723.581,584.804)QPointF(747.622,576.588)QPointF(747.899,571.431)QPointF(766.482,568.15)QPointF(774.357,562.801)QPointF(767.935,563.462)QPointF(785.301,560.92)QPointF(785.337,560.92)QPointF(767.975,563.462)QPointF(774.401,562.801)QPointF(766.53,568.15)QPointF(747.953,571.431)QPointF(747.682,576.588)QPointF(723.649,584.804)QPointF(713.132,591.03)QPointF(687.833,601.825)QPointF(655.588,615.075)QPointF(606.266,630.116)QPointF(540.581,649.28)QPointF(442.195,674.219)QPointF(319.055,707.915)QPointF(155.056,751.289)QPointF(-78.0515,813.953)QPointF(-420.96,904.149))
> selfdrive/ui/qt/onroad.cc: draw time: 1.28275 ms  
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 86.5993 100
> selfdrive/ui/qt/onroad.cc: using bad poly
> selfdrive/ui/qt/onroad.cc: is closed: false
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-227.194,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
> selfdrive/ui/qt/onroad.cc: draw time: 54.2867 ms high! 
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 82.4211 100
> selfdrive/ui/qt/onroad.cc: is closed: true
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-226.891,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
> selfdrive/ui/qt/onroad.cc: draw time: 48.1844 ms high! 
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 61.5824 100
> selfdrive/ui/qt/onroad.cc: is closed: true
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-226.891,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
> selfdrive/ui/qt/onroad.cc: draw time: 48.1316 ms high!
> ```

It looks like the polygon was drawn using software simulation instead of hardware acceleration

> Certain QPolygonFs cause these high frame times consistently, for example this waits until a high draw time, then keeps drawing it. It then is fixed if you convert it to an integer polygon with .toPolygon()
> 
> ```
> selfdrive/ui/ui.cc: max_distance orig, new: 73.8285 100
> selfdrive/ui/qt/onroad.cc: is closed: false
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-451.948,899.583)QPointF(-99.3411,810.203)QPointF(146.246,748.129)QPointF(319.808,704.831)QPointF(455.683,670.959)QPointF(564.931,645.632)QPointF(635.444,626.562)QPointF(697.379,611.425)QPointF(735.719,598.36)QPointF(768.944,587.918)QPointF(786.11,581.553)QPointF(816.362,573.425)QPointF(821.051,568.269)QPointF(843.466,564.851)QPointF(852.994,559.49)QPointF(848.498,560.107)QPointF(864.489,558.175)QPointF(864.526,558.175)QPointF(848.538,560.107)QPointF(853.038,559.49)QPointF(843.514,564.851)QPointF(821.105,568.269)QPointF(816.422,573.425)QPointF(786.177,581.553)QPointF(769.019,587.918)QPointF(735.806,598.36)QPointF(697.478,611.425)QPointF(635.559,626.562)QPointF(565.065,645.632)QPointF(455.843,670.959)QPointF(320.003,704.831)QPointF(146.485,748.129)QPointF(-99.0377,810.203)QPointF(-451.552,899.583))
> selfdrive/ui/qt/onroad.cc: draw time: 1.18842 ms  
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 100 100
> selfdrive/ui/qt/onroad.cc: is closed: false
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-421.356,904.149)QPointF(-78.3551,813.953)QPointF(154.816,751.289)QPointF(318.861,707.915)QPointF(442.034,674.219)QPointF(540.446,649.28)QPointF(606.151,630.116)QPointF(655.488,615.075)QPointF(687.746,601.825)QPointF(713.056,591.03)QPointF(723.581,584.804)QPointF(747.622,576.588)QPointF(747.899,571.431)QPointF(766.482,568.15)QPointF(774.357,562.801)QPointF(767.935,563.462)QPointF(785.301,560.92)QPointF(785.337,560.92)QPointF(767.975,563.462)QPointF(774.401,562.801)QPointF(766.53,568.15)QPointF(747.953,571.431)QPointF(747.682,576.588)QPointF(723.649,584.804)QPointF(713.132,591.03)QPointF(687.833,601.825)QPointF(655.588,615.075)QPointF(606.266,630.116)QPointF(540.581,649.28)QPointF(442.195,674.219)QPointF(319.055,707.915)QPointF(155.056,751.289)QPointF(-78.0515,813.953)QPointF(-420.96,904.149))
> selfdrive/ui/qt/onroad.cc: draw time: 1.28275 ms  
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 86.5993 100
> selfdrive/ui/qt/onroad.cc: using bad poly
> selfdrive/ui/qt/onroad.cc: is closed: false
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-227.194,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
> selfdrive/ui/qt/onroad.cc: draw time: 54.2867 ms high! 
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 82.4211 100
> selfdrive/ui/qt/onroad.cc: is closed: true
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-226.891,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
> selfdrive/ui/qt/onroad.cc: draw time: 48.1844 ms high! 
> 
> selfdrive/ui/ui.cc: max_distance orig, new: 61.5824 100
> selfdrive/ui/qt/onroad.cc: is closed: true
> selfdrive/ui/qt/onroad.cc: QPolygonF(QPointF(-226.891,810.096)QPointF(8.98217,748.223)QPointF(161.815,705.726)QPointF(275.524,672.734)QPointF(359.59,648.144)QPointF(400.718,629.924)QPointF(440.32,615.716)QPointF(445.962,602.879)QPointF(453.59,593.545)QPointF(447.754,587.792)QPointF(452.235,580.514)QPointF(442.916,575.783)QPointF(446.003,573.375)QPointF(445.972,568.046)QPointF(427.373,568.612)QPointF(445.016,567.363)QPointF(445.053,567.363)QPointF(427.413,568.612)QPointF(446.017,568.046)QPointF(446.051,573.375)QPointF(442.969,575.783)QPointF(452.294,580.514)QPointF(447.821,587.792)QPointF(453.666,593.545)QPointF(446.048,602.879)QPointF(440.419,615.716)QPointF(400.833,629.924)QPointF(359.725,648.144)QPointF(275.684,672.734)QPointF(162.009,705.726)QPointF(9.22202,748.223)QPointF(-226.891,810.096))
> selfdrive/ui/qt/onroad.cc: draw time: 48.1316 ms high!
> ```


If it takes tat much time, the polygon should have been drawn using software simulation instead of hardware acceleration

**Comment by deanlee** [comment]
> 

If it takes tat much time, the polygon should have been drawn using software simulation instead of hardware acceleration

**Comment by sshane** [comment]
It's only certain polygons, why would that happen?

**Comment by deanlee** [comment]
This will happen when qt's opengl PaintEnginee doesn't know how to close the points. In this case qt will use Qpaintpath to do a software simulation rending.

**Comment by deanlee** [comment]
Is there a significant difference if you replace drawPolygon with drawConvexPolygon?

**Comment by sshane** [comment]
Yes, no spikes, but leads to undefined behavior.

(just drawing one)

image

**Comment by deanlee** [comment]
So it looks like we can solve this issue by making the first point in polygon implicitly connected to the last point. this would satisfy the conditions for hardware acceleration.

**Comment by sshane** [comment]
So I profiled both a good section of the route (with thick lane line) and compared it to the bad section (with thin line).

This is inside QPaintEngineEx::draw(), which up till here seemed similar (no extra function calls). The ir per call is massively different here:
Good:
image

Bad:
image

**Comment by deanlee** [comment]
drawConvexPolygon will set mode to ConvexMode https://github.com/qt/qt/blob/0a2f2382541424726168804be2c90b91381608c6/src/opengl/qpaintengine_opengl.cpp#L3870, it is guaranteed to use hardware acceleration:


```
f (mode == ConvexMode && !d->high_quality_antialiasing) {
            //### resolving on polygon from points isn't correct
            d->setGradientOps(d->cbrush, bounds);

            const qreal *vertexArray = reinterpret_cast<const qreal*>(&points[0]);

            if (sizeof(qreal) == sizeof(double)) {
                Q_ASSERT(sizeof(QPointF) == 16);
                glVertexPointer(2, GL_DOUBLE, 0, vertexArray);
            }
            else {
                Q_ASSERT(sizeof(QPointF) == 8);
                glVertexPointer(2, GL_FLOAT, 0, vertexArray);
            }

            glEnableClientState(GL_VERTEX_ARRAY);
            glDrawArrays(GL_TRIANGLE_FAN, 0, pointCount);
            glDisableClientState(GL_VERTEX_ARRAY);
        } 
```

**Comment by sshane** [comment]
But lane lines are not convex, right?

**Comment by sshane** [comment]
Yeah think you're right, qTriangulate must be on the CPU. If you enable stencil buffering, it uses that instead of qTriangulate and the frame times are much more consistent:

https://github.com/qt/qt/blob/0a2f2382541424726168804be2c90b91381608c6/src/opengl/gl2paintengineex/qpaintengineex_opengl2.cpp#L942-L966

**Comment by sshane** [comment]
And here's running on that C3 in the testing closet! Switched to 1 bit buffer size half way through

image

File: selfdrive/ui/ui.cc (hunk 1 above)
Comment (sshane) [reply selfdrive/ui/ui.cc]: This likely only worked because it likely reduced triangles near the end of the ll

**Comment by deanlee** [comment]
WOW! so it has been resolved by enabling stencil buffering? that's awesome!!!

**Review by adeebshihadeh — APPROVED** [review]

**Comment by sshane** [comment]
One final test on two different devices. One is on jenkins and is positioned to generate lane lines with many intersections and polygon is very thin at the end (causing qTriangulate to return visual artifacts and take longer to calculate). The other is at my desk and lane lines are so close to 0 prob you can't see them.

Jenkins device:

image

Desk device:

image

**Comment by sshane** [comment]
@deanlee you think this is related to need to round the hue in the path? If you can reproduce that, want to test before and after this PR with the rounding removed?

**Pull request merged**